fix: windows use powershell to extract zip file

jacobtread opened the pull request:
## Description

For windows platforms the unzip utility is not on systems by default, this changes it so that on windows powershell is used in combination with Expand-Archive to extract the zip file.

## Changes
- Extend "Package" to take in the platform
- Add case in zip extraction for the windows platform to use powershell Expand-Archive instead of unzip

## Related Issues
- Closes #2004

mistydemeo commented:
Thank you very much! I appreciate this! I'll give it a test and see how it works.

It looks like `prettier` is flagging a couple of requested formatting changes. Are you up for tweaking that? I can take care of it if you prefer.

The snapshots we use for tests will need to be updated with these changes. I can do that for you if you'd like, or you can do it by following these steps:

* Install [cargo insta](https://insta.rs/docs/cli/)
* Run the tests with `cargo test`
* Run `cargo insta accept`
* Commit and push the changes

jacobtread commented:
Hi, I have updated the formatting :) I will need you to please handle the test snapshots updating I think If I run them on my device it will cause lots of problems, as running the tests on my machine ends up failing a majority of unrelated integration and snapshot tests (I'd assume because I'm using windows and the tests don't take that into account) thank you!

mistydemeo commented:
Thanks! Looks like one snapshot didn't take, so I've pushed a commit with that.

mistydemeo commented:
Looking good! I'll just do some testing on my Windows machine later to make sure all works as I expect, but I think this should be good.

mistydemeo commented on cargo-dist/templates/installer/npm/binary-install.js:
```diff
@@ -122,12 +124,25 @@ class Package {
+                    "-NoProfile",
+                    "-NonInteractive",
+                    "-Command",
+                    `Expand-Archive -LiteralPath '${tempFile}' -DestinationPath '${this.installDirectory}' -Force`,
```
Just had a realization. Does this introduce the possibility of shell injection? Previously that wasn't possible because `spawnSync` is being run with `shell=false`, but in this version the unzip command is being passed to powershell as a single command string. Could mean that install directories or tempfile names with powershell parameters in them could do things here they don't in the other paths.

jacobtread replied:
Yes I think you are right there, I will have a look into other possible ways of executing the command

jacobtread commented:
> Sorry for the delay! I’ve just been very busy lately, and hope to have time to integrate and release this soon.

It's no worries!

I have modified the command to use args to prevent shell injection in the install dir / temp file name. 

I've updated the snapshots that would run for me the "axolotlsay_basic" test is failing on my machine due to "failed to generate an OmniBOR Artifact ID (status: exit code: 2)" so If you could please run that on your machine for me if I've missed any snapshots 😄

StartAutomating commented on cargo-dist/templates/installer/npm/binary-install.js:
```diff
@@ -122,12 +124,27 @@ class Package {
+                    "-NoProfile",
+                    "-NonInteractive",
+                    "-Command",
+                    "& { Expand-Archive -LiteralPath $args[0] -DestinationPath $args[1] -Force }",
```
Expand-Archive is safe, but this technique of embedding isn't always so secure.

Essentially, because you're passing down an unquoted string in thru, it could theoretically get expanded into something nasty.  In practice, you'll probably be fine if the inputs are sanitized first (say, by confirming the file exists).

I'd personally feel more comfortable if the command were something like:

~~~PowerShell
& {param([string]$LiteralPath, [string]$DestinationPath) Expand-Archive @psboundParameters } 
~~~

This would force each parameter to be a single string, and pass the parameters directly down to Expand-Path.

mistydemeo replied:
Great, thank you!

mistydemeo commented:
I'll try to get a new release out soon.
